Schema Validation Issues

Schema-related tests fail due to missing or incorrectly structured data in the user_base_data, user_create_data, and user_response_data fixtures. Additionally, some validation errors occur with UUID parsing.

Affected Tests:

test_user_base_valid (KeyError: 'nickname')
test_user_create_valid (KeyError: 'nickname')
test_user_update_valid (KeyError: 'first_name')
test_user_response_valid (UUID validation error)
test_login_request_valid (Field required [type=missing])

**_Add below fixtures to conftest.py:_**

```
# Fixtures for common test data
@pytest.fixture
def user_base_data():
    return {
        "username": "john_doe_123",
        "email": "john.doe@example.com",
        "full_name": "John Doe",
        "nickname": "johnjohn",
        "bio": "I am a software engineer with over 5 years of experience.",
        "profile_picture_url": "https://example.com/profile_pictures/john_doe.jpg"
    }


# Tests for LoginRequest
def test_login_request_valid(login_request_data):
    login = LoginRequest(**login_request_data)
    assert login.email == login_request_data["email"]
    assert login.password == login_request_data["password"]
```

Remaining Issue:
```
FAILED tests/test_schemas/test_user_schemas.py::test_user_response_valid - pydantic_core._pydantic_core.ValidationError: 1 validation error for UserResponse

Error:
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for UserResponse
E       id
E         Input should be a valid UUID, invalid character: expected an optional prefix of `urn:uuid:` followed by [0-9a-fA-F-], found `u` at 1 [type=uuid_parsing, input_value='unique-id-string', input_type=str]
E           For further information visit https://errors.pydantic.dev/2.6/v/uuid_parsing

tests/test_schemas/test_user_schemas.py:27: ValidationError
```
Fix: Modify id  in the user_response_fixture to return uuid.uuid4() instead of str(uuid.uuid4()) as the model UserResponse contains id of type UUID in conftest.py

AttributeError: type object 'datetime.datetime' has no attribute 'datetime'

Fix: Add in missing fields  created_at, updated_at, last login_at in the UserResponse model in user_schemas.py and, in the conftest.py, add import statement from datetime import datetime. Also need to set created_at to datetime.now() in the user_response_data fixture.

    def test_login_request_valid(login_request_data):
>       login = LoginRequest(**login_request_data)
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for LoginRequest
E       email
E         Field required [type=missing, input_value={'username': 'john_doe_12...': 'SecurePassword123!'}, input_type=dict]
E           For further information visit https://errors.pydantic.dev/2.6/v/missing

tests/test_schemas/test_user_schemas.py:35: ValidationError

Fix: Change email to username in LoginRequest model in user_schemas.py and also make change to test_login_request_valid assertion statement where login.username==login_request_data["username"].

FAILED tests/test_api/test_users_api.py::test_update_user_github - AttributeError: 'str' object has no attribute 'values'

Fix: a check to see if values in check_at_least_one_value validator is a dict type. if not, then we return the singular string. This is needed because a string does not have a .values() attribute whereas a dict does.

**Steps Attempted**
1. Updating test cases;
2. Updating some outdated pydantic functions and keywords so it matches what we need;
3. Review codes and understand how they interacting with each other, checking specific endpoints to ensure it meets the expected behavior. For example, test `test_update_user_email_access_allowed` failed

--- a/app/schemas/user_schemas.py
+++ b/app/schemas/user_schemas.py
@@ -1,7 +1,8 @@
 from builtins import ValueError, any, bool, str
-from pydantic import BaseModel, EmailStr, Field, validator, root_validator
+from pydantic import BaseModel, EmailStr, Field, field_validator
 from typing import Optional, List
 from datetime import datetime
+from app.models.user_model import UserRole
 from enum import Enum
 import uuid
 import re
@@ -32,7 +33,12 @@ class UserBase(BaseModel):
     linkedin_profile_url: Optional[str] =Field(None, example="https://linkedin.com/in/johndoe")
     github_profile_url: Optional[str] = Field(None, example="https://github.com/johndoe")
 
-    _validate_urls = validator('profile_picture_url', 'linkedin_profile_url', 'github_profile_url', pre=True, allow_reuse=True)(validate_url)
+    # Corrected field_validator usage
+    @field_validator('profile_picture_url', 'linkedin_profile_url', 'github_profile_url', mode='before')
+    def validate_urls(cls, value):
+        if value:
+            validate_url(value)  # Assuming validate_url raises exceptions if the URL is invalid
+        return value
  
     class Config:
         from_attributes = True
@@ -51,9 +57,9 @@ class UserUpdate(UserBase):
     linkedin_profile_url: Optional[str] =Field(None, example="https://linkedin.com/in/johndoe")
     github_profile_url: Optional[str] = Field(None, example="https://github.com/johndoe")
 
-    @root_validator(pre=True)
+    @field_validator('*', mode='before')
     def check_at_least_one_value(cls, values):
-        if not any(values.values()):
+        if isinstance(values,dict) and not any(values.values()):
             raise ValueError("At least one field must be provided for update")
         return values
 
@@ -61,12 +67,14 @@ class UserResponse(UserBase):
     id: uuid.UUID = Field(..., example=uuid.uuid4())
     role: UserRole = Field(default=UserRole.AUTHENTICATED, example="AUTHENTICATED")
     email: EmailStr = Field(..., example="john.doe@example.com")
-    nickname: Optional[str] = Field(None, min_length=3, pattern=r'^[\w-]+$', example=generate_nickname())    
-    role: UserRole = Field(default=UserRole.AUTHENTICATED, example="AUTHENTICATED")
+    nickname: Optional[str] = Field(None, min_length=3, pattern=r'^[\w-]+$', example=generate_nickname())
     is_professional: Optional[bool] = Field(default=False, example=True)
+    created_at: Optional[datetime]
+    updated_at: Optional[datetime]
+    last_login_at: Optional[datetime]
 
 class LoginRequest(BaseModel):
-    email: str = Field(..., example="john.doe@example.com")
+    username: str = Field(..., example="john.doe@example.com")
     password: str = Field(..., example="Secure*1234")
 
 class ErrorResponse(BaseModel):

--- a/app/utils/smtp_connection.py
+++ b/app/utils/smtp_connection.py
@@ -22,12 +22,7 @@ class SMTPClient:
             message.attach(MIMEText(html_content, 'html'))
 
             with smtplib.SMTP(self.server, self.port) as server:
-                server.starttls()  # Use TLS
-                print(f"Username: {self.username}")
-                print(f"Password': {self.password}")
-                print(f"HTML CONTENT: {html_content}")
-                print(f"Message: {message.as_string()}")
-                      
+                server.starttls()  # Use TLS      
                 server.login(self.username, self.password)
                 server.sendmail(self.username, recipient, message.as_string())
             logging.info(f"Email sent to {recipient}")

--- a/readme.md
+++ b/readme.md
@@ -6,7 +6,10 @@ This repository hosts the source code for the Event Manager project, a web appli
 This document details the progress made on the project, highlights resolved and unresolved issues, and provides insights into the learning process during the development journey.
 
 ## Github Links
-**Closed Issues**: Links\
+**Closed Issues**: 
+1. [Issue#1](https://github.com/Livia-1212/event_manager/issues/1)
+3. [Issue#3](https://github.com/Livia-1212/event_manager/issues/3)
+
 **Dockerhub Image**: Dockerhub Image
 
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -21,6 +21,7 @@ from uuid import uuid4
 
 # Third-party imports
 import pytest
+import uuid
 from httpx import AsyncClient
 from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
 from sqlalchemy.orm import sessionmaker, scoped_session
@@ -35,6 +36,7 @@ from app.utils.security import hash_password
 from app.utils.template_manager import TemplateManager
 from app.services.email_service import EmailService
 from app.services.jwt_service import create_access_token
+from datetime import datetime
 
 fake = Faker()
 
@@ -250,13 +252,15 @@ def user_update_data():
 @pytest.fixture
 def user_response_data():
     return {
-        "id": "unique-id-string",
-        "username": "testuser",
-        "email": "test@example.com",
-        "last_login_at": datetime.now(),
+        "id": uuid.uuid4(),
+        "email": "john.doe@example.com",
+        "full_name": "John Doe",
+        "nickname": "johnjohn",
+        "role": "AUTHENTICATED",
+        "is_professional": True,
         "created_at": datetime.now(),
         "updated_at": datetime.now(),
-        "links": []
+        "last_login_at": datetime.now(),
     }
 
 @pytest.fixture

--- a/tests/test_schemas/test_user_schemas.py
+++ b/tests/test_schemas/test_user_schemas.py
@@ -31,7 +31,7 @@ def test_user_response_valid(user_response_data):
 # Tests for LoginRequest
 def test_login_request_valid(login_request_data):
     login = LoginRequest(**login_request_data)
-    assert login.email == login_request_data["email"]
+    assert login.username == login_request_data["username"]
     assert login.password == login_request_data["password"]
 
 # Parametrized tests for nickname and email validation

